contrib/src/main/java/gov/nasa/jpl/aerie/contrib/models/SampledResource.java

```diff
@@ -1,35 +1,68 @@
 package gov.nasa.jpl.aerie.contrib.models;
 
 import gov.nasa.jpl.aerie.merlin.framework.resources.discrete.DiscreteResource;
+import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
 
 import java.util.Objects;
 import java.util.function.Supplier;
 import java.util.function.UnaryOperator;
 
 import static gov.nasa.jpl.aerie.merlin.framework.ModelActions.*;
-import static gov.nasa.jpl.aerie.merlin.protocol.types.Duration.SECOND;
 
+/**
+ * Simple resource that samples arbitrarily many existing resources/values at a specified rate (default rate is once per
+ * second).
+ */
 public class SampledResource<T> implements DiscreteResource<T> {
   private final Register<T> result;
   private final Supplier<T> sampler;
+  private double rate;
 
+  /**
+   * Constructor that does not require caller to specify a rate and therefore assumes a sample
+   * rate of 1 sample per second.
+   */
   public SampledResource(final Supplier<T> sampler, final UnaryOperator<T> duplicator) {
     this.result = Register.create(sampler.get(), duplicator);
     this.sampler = Objects.requireNonNull(sampler);
+    this.rate = 1.0;
+    spawn(this::takeSamples);
+  }
 
+  /**
+   * Constructor that requires caller to specify an initial sample rate
+   */
+  public SampledResource(final Supplier<T> sampler, final UnaryOperator<T> duplicator, final double rate) {
+    this.result = Register.create(sampler.get(), duplicator);
+    this.sampler = Objects.requireNonNull(sampler);
+    this.rate = rate;
     spawn(this::takeSamples);
   }
 
+  /**
+   * Method that samples the supplied resource at the current specified rate
+   */
   private void takeSamples() {
     while (true) {
       var sample = sampler.get();
       if (!result.get().equals(sample)) {
         result.set(sample);
       }
-      delay(1, SECOND);
+      delay( Duration.roundNearest(rate, Duration.SECONDS) );
     }
   }
 
+  /**
+   * Get current sample rate
+   */
+  public double getRate() { return rate; }
+
+  /**
+   * Method to adjust the specified rate of sampling. Note if takeSamples() is currently waiting, the
+   * new rate will not take effect until after the current wait cycle.
+   */
+  public void setRate(final double newRate ) { rate = newRate; }
```

Comment:
I'm pondering whether `rate` should be tracked in a Cell, or if it's okay as a field 🤔 To err on the safe side I would recommend wrapping `rate` in a `Register`. 

The hazard here is that if Task A and Task B are executing concurrently - and Task A calls `setRate`, while Task B calls `getRate`, we can accidentally create a causal dependency that is invisible to Aerie.

However, if by policy this SampledResource's rate is accessed by a single daemon task, you may be okay with it as is. Pinging @Twisol in case he'd like to weigh in

Reply:
Pardon my ignorance, but what are the implications of this causal dependency on the Aerie system? I'll make the update nonetheless.

Reply:
Yes, in general all mutable state in a Merlin model needs to be mediated by a Cell managed by Merlin. There are a couple of reasons for this, but a big one is that various parts of a simulated task might be repeated for any number of reasons, and we need to idempotize any changes to state so that they only take effect the first time they occur.

Probably the most visible and accessible way this can arise in the current system is through the use of `ReplayingTask`. The default mode of executing tasks wraps each task in its thread; this is called a `ThreadedTask`. However, the overhead of context-switching, especially for tasks that only execute for a brief amount of time before yielding control again, adds up very quickly. For many kinds of task, using a replaying strategy is faster: when a task yields, it actually throws an exception to bail out; and when the task resumes, we call into it from the top again and run it forward until it gets back to where it was before, skipping any effects it performed along the way. It might seem wasteful to re-run the same code, but for "most" tasks, the overhead of rerunning the logic is drastically outweighed by the overhead of context-switching.

There are more "correctness"-oriented reasons, too: any resource exposed from a model is polled by the simulation host only when we detect changes to the Cells it depends on. If a resource is computed from unmanaged state, either in whole or in part, we cannot record all of the changes that ought to be reported to planners. We could poll, instead, but the specter of performance raises its head again: polling is extremely expensive overall, and checking a resource only when we *know* we need to reduces our overhead considerably.

Our approach also means that for many resources, sampling rates don't need to be managed explicitly; and for those that do, there is a path involving the `Cell.getExpiryRate` function to tell the simulation host when the last read should be considered invalidated by default, such as for complex polynomial resources: since we can only currently report linear segments, at some point the last linear segment will fall too far out of line with the actual curve, so we can say that the segment "expires" after a certain amount of time. This is also more performant than either `ThreadedTask` or `ReplayingTask`: the former incurs a context switch, and the latter incurs at least the cost of idempotized effect history. Resource expiration can be handled directly from the simulation host, keeping the task execution machinery off the hot path altogether.

Lastly, and most future-oriented, incremental resimulation strategies uniformly depend on us either (or both) having a complete view on the state used by a model (so that we can snapshot and restore it as needed), or be able to re-run only the tasks which depend on data that has changed -- if some model state is not affected by a change in a plan, then tasks that depend on it may not need to be re-run at all. If unmanaged mutable state exists, then it is very easy to end up in a situation where the behavior of the model doesn't line up with what it would have done in a fresh, non-incremental run. This kind of divergence can completely neuter the benefit of incremental resimulation -- it doesn't matter how fast it is if it's wrong!

Reply:
Thanks for the detailed explanation!